File: app/src/main/java/com/example/finanzas/data/api/ImportService.java
package com.example.finanzas.data.api;

import android.content.Context;

import com.example.finanzas.data.local.LocalRepository;
import com.example.finanzas.data.model.ImportJob;
import com.example.finanzas.data.model.ImportRule;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;

public class ImportService {

    public interface ListCallback {
        void onSuccess(List<ImportJob> items);
        void onError();
    }

    public interface RulesCallback {
        void onSuccess(List<ImportRule> rules);
        void onError();
    }

    public interface CreateCallback {
        void onSuccess(int importId);
        void onError();
    }

    public interface SimpleCallback {
        void onSuccess();
        void onError();
    }

    public interface ProcessCallback {
        void onSuccess(JSONObject response);
        void onError();
    }

    public static void list(Context ctx, ListCallback cb) {
        try {
            cb.onSuccess(LocalRepository.getInstance(ctx).listImports());
        } catch (Exception e) {
            cb.onError();
        }
    }

    public static void listRules(Context ctx, RulesCallback cb) {
        try {
            cb.onSuccess(LocalRepository.getInstance(ctx).listImportRules());
        } catch (Exception e) {
            cb.onError();
        }
    }

    public static void create(Context ctx, String tipo, String nombre, JSONArray lineas, CreateCallback cb) {
        try {
            int id = LocalRepository.getInstance(ctx).createImport(nombre, tipo, lineas != null ? lineas.toString() : "[]");
            cb.onSuccess(id);
        } catch (Exception e) {
            cb.onError();
        }
    }

    public static void saveRule(Context ctx, ImportRule rule, SimpleCallback cb) {
        try {
            LocalRepository.getInstance(ctx).saveImportRule(rule);
            cb.onSuccess();
        } catch (Exception e) {
            cb.onError();
        }
    }

    public static void deleteRule(Context ctx, int ruleId, SimpleCallback cb) {
        try {
            LocalRepository.getInstance(ctx).deleteImportRule(ruleId);
            cb.onSuccess();
        } catch (Exception e) {
            cb.onError();
        }
    }

    public static void process(Context ctx, int importId, ProcessCallback cb) {
        try {
            JSONObject result = LocalRepository.getInstance(ctx).processImport(importId);
            cb.onSuccess(result);
        } catch (Exception e) {
            cb.onError();
        }
    }
}

File: app/src/main/java/com/example/finanzas/data/api/ImportService.kt
package com.example.finanzas.data.api

import android.content.Context
import com.example.finanzas.data.local.LocalRepository
import com.example.finanzas.data.model.ImportJob
import com.example.finanzas.data.model.ImportRule
import org.json.JSONArray
import org.json.JSONObject

object ImportService {

    interface ListCallback {
        fun onSuccess(items: List<ImportJob>)
        fun onError()
    }

    interface RulesCallback {
        fun onSuccess(rules: List<ImportRule>)
        fun onError()
    }

    interface CreateCallback {
        fun onSuccess(importId: Int)
        fun onError()
    }

    interface SimpleCallback {
        fun onSuccess()
        fun onError()
    }

    interface ProcessCallback {
        fun onSuccess(response: JSONObject)
        fun onError()
    }

    @JvmStatic
    fun list(ctx: Context, cb: ListCallback) {
        DbCoroutine.io(
            block = { LocalRepository.getInstance(ctx).listImports() },
            onSuccess = cb::onSuccess,
            onError = cb::onError
        )
    }

    @JvmStatic
    fun listRules(ctx: Context, cb: RulesCallback) {
        DbCoroutine.io(
            block = { LocalRepository.getInstance(ctx).listImportRules() },
            onSuccess = cb::onSuccess,
            onError = cb::onError
        )
    }

    @JvmStatic
    fun create(ctx: Context, tipo: String, nombre: String, lineas: JSONArray?, cb: CreateCallback) {
        DbCoroutine.io(
            block = {
                LocalRepository.getInstance(ctx).createImport(
                    nombre,
                    tipo,
                    lineas?.toString() ?: "[]"
                )
            },
            onSuccess = cb::onSuccess,
            onError = cb::onError
        )
    }

    @JvmStatic
    fun saveRule(ctx: Context, rule: ImportRule, cb: SimpleCallback) {
        DbCoroutine.io(
            block = {
                LocalRepository.getInstance(ctx).saveImportRule(rule)
                Unit
            },
            onSuccess = { cb.onSuccess() },
            onError = cb::onError
        )
    }

    @JvmStatic
    fun deleteRule(ctx: Context, ruleId: Int, cb: SimpleCallback) {
        DbCoroutine.io(
            block = {
                LocalRepository.getInstance(ctx).deleteImportRule(ruleId)
                Unit
            },
            onSuccess = { cb.onSuccess() },
            onError = cb::onError
        )
    }

    @JvmStatic
    fun process(ctx: Context, importId: Int, cb: ProcessCallback) {
        DbCoroutine.io(
            block = { LocalRepository.getInstance(ctx).processImport(importId) },
            onSuccess = cb::onSuccess,
            onError = cb::onError
        )
    }
}
